### PR TITLE
Cache interface improvement

### PR DESCRIPTION
**Pull request by Tuanhao**

### CODE REVIEW

**Comment by Tuanhao** [comment]
@ttjiaa PTAL

File: jwk_client.go [hunk 1]
```diff
@@ -11,8 +11,7 @@ import (
 )
 
 var (
-	ErrInvalidContentType = errors.New("Should have a JSON content type for JWKS endpoint.")
-	ErrNoKeyFound         = errors.New("No Keys has been found")
+	ErrInvalidContentType = errors.New("Should have a JSON content type for JWKS endpoint")
 	ErrInvalidAlgorithm   = errors.New("Algorithm is invalid")
```
Comment (ttjiaa) [inline jwk_client.go]: make error msgs lowercase

File: jwk_client.go [hunk 2]
```diff
@@ -67,20 +66,17 @@ func (j *JWKClient) GetKey(ID string) (jose.JSONWebKey, error) {
...
-			if success {
-				return addedKey, nil
-			}
-			return jose.JSONWebKey{}, ErrNoKeyFound
+			addedKey, err := j.keyCacher.Add(ID, keys)
+			return addedKey, err
 		}
 	}
-	return searchedKey, nil
+	return searchedKey, err
```
Comment (ttjiaa) [inline jwk_client.go]: nil

File: jwk_client_test.go [hunk 3]
```diff
@@ -28,23 +28,23 @@ func newMockKeyCacher(get bool, add bool, key jose.JSONWebKey, keyID string) *mo
 	}
 }
 
-func (mockKC *mockKeyCacher) Get(keyID string) (jose.JSONWebKey, bool) {
+func (mockKC *mockKeyCacher) Get(keyID string) (jose.JSONWebKey, error) {
 	if mockKC.get {
```
Comment (ttjiaa) [inline jwk_client_test.go]: change from `get` as `bool` to `getError` `error` for mock key cacher, same for `add`

File: jwk_client.go [hunk 4]
```diff
@@ -67,20 +68,20 @@ func (j *JWKClient) GetKey(ID string) (jose.JSONWebKey, error) {
 	j.mu.Lock()
 	defer j.mu.Unlock()
 
-	searchedKey, exist := j.keyCacher.Get(ID)
+	searchedKey, err := j.keyCacher.Get(ID)
 
-	if !exist {
+	if searchedKey == nil {
```
Comment (ttjiaa) [inline jwk_client.go]: if err

File: common_test.go [hunk 5]
```diff
@@ -109,3 +113,30 @@ func getTestTokenWithKid(audience []string, issuer string, expTime time.Time, al
 	}
 	return raw
 }
+
+func genNewServer(genJWKS bool) (JWKClientOptions, string, string, error) {
```
Comment (ttjiaa) [inline common_test.go]: `genNewTestServer`

File: jwk_client_test.go [hunk 6]
```diff
@@ -127,32 +85,17 @@ func TestJWKDownloadKeyNoKeys(t *testing.T) {
...
-	tokenES384 := getTestTokenWithKid(defaultAudience, defaultIssuer, time.Now().Add(24*time.Hour), jose.ES384, jsonWebKeyES384, "keyES384")
-
-	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		w.Header().Set("Content-Type", "application/json")
-		fmt.Fprintln(w, string(value))
-	}))
-	opts := JWKClientOptions{URI: ts.URL}
 	client := NewJWKClient(opts, nil)
 
+	tokenES385 := getTestTokenWithKid(defaultAudience, defaultIssuer, time.Now().Add(24*time.Hour), jose.ES384, genECDSAJWK(jose.ES384, "keyES385"), "keyES385")
```
Comment (ttjiaa) [inline jwk_client_test.go]: call this something like `invalidToken`

File: key_cacher.go [hunk 7]
```diff
@@ -69,15 +69,14 @@ func (mkc *memoryKeyCacher) Add(keyID string, downloadedKeys []jose.JSONWebKey)
...
 	}
-	if mkc.size != -1 {
-		mkc.entries[addingKey.KeyID] = keyCacherEntry{
-			addedAt:    time.Now(),
-			JSONWebKey: addingKey,
-		}
-		handleOverflow(mkc)
-		return &addingKey, nil
-	}
 	if addingKey.KeyID != "" {
```
Comment (ttjiaa) [inline key_cacher.go]: `if addingKey == jose.JSONWebKey{}`

File: key_cacher.go [hunk 8]
```diff
@@ -39,7 +39,7 @@ func NewMemoryKeyCacher(maxAge time.Duration, size int) KeyCacher {
 func newMemoryPersistentKeyCacher() KeyCacher {
 	return &memoryKeyCacher{
 		entries: map[string]keyCacherEntry{},
-		maxAge:  -1, //time.Time
+		maxAge:  time.Duration(-1),
```
Comment (ttjiaa) [inline key_cacher.go]:
```
var (
    MaxAgeNoCheck := time.Duration(-1)
    MaxSizeNoCheck := -1
)

File: jwk_client.go [hunk 9]
```diff
@@ -68,7 +69,7 @@ func (j *JWKClient) GetKey(ID string) (jose.JSONWebKey, error) {
 
 	searchedKey, err := j.keyCacher.Get(ID)
 	if err != nil {
-		//log err
+		log.Println("Error for getting a key", err)
```
Comment (ttjiaa) [inline jwk_client.go]: log.Debug

File: key_cacher.go [hunk 10]
```diff
@@ -39,15 +41,15 @@ func NewMemoryKeyCacher(maxAge time.Duration, size int) KeyCacher {
 func newMemoryPersistentKeyCacher() KeyCacher {
 	return &memoryKeyCacher{
 		entries: map[string]keyCacherEntry{},
-		maxAge:  time.Duration(-1),
-		size:    -1,
+		maxAge:  MaxAgeNoCheck,
+		size:    SizeNoCheck,
```
Comment (ttjiaa) [inline key_cacher.go]: maxSize

File: key_cacher.go [hunk 11]
```diff
@@ -39,15 +41,15 @@ func NewMemoryKeyCacher(maxAge time.Duration, size int) KeyCacher {
...
 	}
 }
 
 func (mkc *memoryKeyCacher) Get(keyID string) (*jose.JSONWebKey, error) {
 	searchKey, ok := mkc.entries[keyID]
 	if ok {
-		if mkc.size == -1 || !isExpired(mkc, keyID) {
+		if mkc.maxAge == time.Duration(-1) || !isExpired(mkc, keyID) {
```
Comment (ttjiaa) [inline key_cacher.go]: `if mkc.maxAge == MaxAgeNoCheck `

File: README.md [hunk 12]
```diff
@@ -99,6 +99,26 @@ if err != nil {
     fmt.Println("Token is not valid:", token)
 }
 ```
+## Support key cacher interface
```
Comment (seriousben) [inline README.md]: Are all the examples in the README valid?

File: README.md [hunk 13]
```diff
@@ -99,6 +99,26 @@ if err != nil {
...
+### Persistent key cacher:
+
+``` func NewJWKClient(options JWKClientOptions, extractor RequestTokenExtractor) *JWKClient```
```
Comment (seriousben) [inline README.md]: Might be more aligned with the rest of README to actually put full examples here.

File: README.md [hunk 14]
```diff
@@ -99,6 +99,26 @@ if err != nil {
...
+There is a bulit-in in-memory key cacher that supports max age and max size (capacity), and it can be initialized through:
+
+```func NewMemoryKeyCacher(maxAge time.Duration, maxSize int) KeyCacher```
+
+*No caching can be obtained by passing 0 to max age and max size in above function*
+
+### Custom key cacher: 
+
+A custom key cacher interface can be passed and used by the client. It just needs to support two main functions: 
+- Get: ```Get(keyID string) (*jose.JSONWebKey, error)```
```
Comment (seriousben) [inline README.md]: go docs are already available for this kind of documentation.

File: jwk_client.go [hunk 15]
```diff
@@ -38,25 +37,25 @@ func NewJWKClient(options JWKClientOptions, extractor RequestTokenExtractor) *JW
...
 	return &JWKClient{
-		keyCacher: kc,
+		keyCacher: keyCacher,
 		options:   options,
 		extractor: extractor,
 	}
 }
 
-func NewJWKClientWithCustomCacher(options JWKClientOptions, extractor RequestTokenExtractor, kc KeyCacher) *JWKClient {
+func NewJWKClientWithCustomCacher(options JWKClientOptions, extractor RequestTokenExtractor, keyCacher KeyCacher) *JWKClient {
```
Comment (seriousben) [inline jwk_client.go]: I would rename `WithCustomCacher` to `WithCache`

File: jwk_client.go [hunk 16]
```diff
@@ -38,25 +37,25 @@ func NewJWKClient(options JWKClientOptions, extractor RequestTokenExtractor) *JW
 		extractor = RequestTokenExtractorFunc(FromHeader)
```
Comment (seriousben) [inline jwk_client.go]: The new `NewJWKClient`, should just use `NewJWKClientWithCache by passing nil`

Comment (seriousben) [inline jwk_client.go]: Can you also add a comment to `NewJWKClient` regarding how the caching works by default.

File: key_cacher.go [hunk 17]
```diff
@@ -1,101 +1,108 @@
...
 
 type keyCacherEntry struct {
 	addedAt time.Time
 	jose.JSONWebKey
 }
 
-func NewMemoryKeyCacher(maxAge int, size int) KeyCacher {
+func NewMemoryKeyCacher(maxAge time.Duration, maxSize int) KeyCacher {
 	return &memoryKeyCacher{
 		entries: map[string]keyCacherEntry{},
```
Comment (seriousben) [inline key_cacher.go]: Should we move the mutex protecting the map in the cacher or not?

File: jwk_client.go [hunk 18]
```diff
@@ -67,20 +66,20 @@ func (j *JWKClient) GetKey(ID string) (jose.JSONWebKey, error) {
 	j.mu.Lock()
```
Comment (seriousben) [inline jwk_client.go:61]: This is interesting, as it serves 3 purposes now. To serializes `GetKey` calls AND it protect accessed to the maps AND it only allows one download keys at a time. I wonder if we could do better for performance.

**Review by seriousben — APPROVED** [review]

**Pull request closed without merging**